pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java

```diff
@@ -115,15 +116,30 @@ public class TableRebalancer {
   private static final Logger LOGGER = LoggerFactory.getLogger(TableRebalancer.class);
   private final HelixManager _helixManager;
   private final HelixDataAccessor _helixDataAccessor;
+  private final TableRebalanceObserver _tableRebalanceObserver;
 
-  public TableRebalancer(HelixManager helixManager) {
+  public TableRebalancer(HelixManager helixManager, @Nullable TableRebalanceObserver tableRebalanceObserver) {
     _helixManager = helixManager;
+    if (tableRebalanceObserver != null) {
+      _tableRebalanceObserver = tableRebalanceObserver;
+    } else {
+      _tableRebalanceObserver = new DefaultTableRebalanceObserver();
+    }
     _helixDataAccessor = helixManager.getHelixDataAccessor();
   }
 
+  public TableRebalancer(HelixManager helixManager) {
+    this(helixManager, null);
+  }
+
+  public static String createUniqueRebalanceJobIdentifier() {
+    return UUID.randomUUID().toString();
+  }
+
   public RebalanceResult rebalance(TableConfig tableConfig, Configuration rebalanceConfig) {
     long startTimeMs = System.currentTimeMillis();
     String tableNameWithType = tableConfig.getTableName();
+    String rebalanceJobId = rebalanceConfig.getString(RebalanceConfigConstants.REBALANCE_ID);
```

Comment:
To be more robust, we can create a unique job id when it is not passed from the rebalance config. Also, with this change, we don't need to modify the code when rebalance is not triggered by the controller (client code doesn't need to be changed as well)

We should also log the rebalance job id in the logger messages for easier tracking